<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>
  Class: FileMaster
  
    &mdash; Documentation by YARD 0.9.28
  
</title>

  <link rel="stylesheet" href="css/style.css" type="text/css" />

  <link rel="stylesheet" href="css/common.css" type="text/css" />

<script type="text/javascript">
  pathId = "FileMaster";
  relpath = '';
</script>


  <script type="text/javascript" charset="utf-8" src="js/jquery.js"></script>

  <script type="text/javascript" charset="utf-8" src="js/app.js"></script>


  </head>
  <body>
    <div class="nav_wrap">
      <iframe id="nav" src="class_list.html?1"></iframe>
      <div id="resizer"></div>
    </div>

    <div id="main" tabindex="-1">
      <div id="header">
        <div id="menu">
  
    <a href="_index.html">Index (F)</a> &raquo;
    
    
    <span class="title">FileMaster</span>
  
</div>

        <div id="search">
  
    <a class="full_list_link" id="class_list_link"
        href="class_list.html">

        <svg width="24" height="24">
          <rect x="0" y="4" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="12" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="20" width="24" height="4" rx="1" ry="1"></rect>
        </svg>
    </a>
  
</div>
        <div class="clear"></div>
      </div>

      <div id="content"><h1>Class: FileMaster
  
  
  
</h1>
<div class="box_info">
  
  <dl>
    <dt>Inherits:</dt>
    <dd>
      <span class="inheritName">Object</span>
      
        <ul class="fullTree">
          <li>Object</li>
          
            <li class="next">FileMaster</li>
          
        </ul>
        <a href="#" class="inheritanceTree">show all</a>
      
    </dd>
  </dl>
  

  
  
  
  
  

  

  
  <dl>
    <dt>Defined in:</dt>
    <dd>FileMaster.rb</dd>
  </dl>
  
</div>

<h2>Overview</h2><div class="docstring">
  <div class="discussion">
    
<h2 id="label-FileMaster">FileMaster</h2>

<p>Class to read file containing the gene list, write the reports files and fetch information froma a URL</p>

<p>Contains the functions to read the file and create a genelist array, to fetch information from API and to create the reports file for genes with no CTTCTT repeats, the GFF3 file with gene relative coordinates and the GFF3 file with chromosome absolute coordinates</p>

<h2 id="label-Summary">Summary</h2>

<p>This class can be use to load the genelist file, to use the fetch function and to write the reports files</p>


  </div>
</div>
<div class="tags">
  

</div>
  
    <h2>
      Constant Summary
      <small><a href="#" class="constants_summary_toggle">collapse</a></small>
    </h2>

    <dl class="constants">
      
        <dt id="genelist-classvariable" class="">@@genelist =
          <div class="docstring">
  <div class="discussion">
    
<p>Class variable [Array&lt;String&gt;] Contains every gene id from the genelist file</p>


  </div>
</div>
<div class="tags">
  

</div>
        </dt>
        <dd><pre class="code"><span class='lbracket'>[</span><span class='rbracket'>]</span></pre></dd>
      
    </dl>
  







  
    <h2>
      Class Method Summary
      <small><a href="#" class="summary_toggle">collapse</a></small>
    </h2>

    <ul class="summary">
      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#fetch-class_method" title="fetch (class method)">.<strong>fetch</strong>(url:, headers: {accept: &quot;*/*&quot;}, user: &quot;&quot;, pass: &quot;&quot;)  &#x21d2; String, false </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Class method fetch to access an URL via code, donated by Mark Wilkinson.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#generate_gff3_report-class_method" title="generate_gff3_report (class method)">.<strong>generate_gff3_report</strong>(seq_objects)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Class method to write GFF3 report with all the CTTCTT regions found in the genes from the list.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#generate_gff3_report_chr-class_method" title="generate_gff3_report_chr (class method)">.<strong>generate_gff3_report_chr</strong>(seq_objects)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Class method to write GFF3 report with all the CTTCTT regions found in the genes from the list.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#generate_no_report-class_method" title="generate_no_report (class method)">.<strong>generate_no_report</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Class method to write report with all the genes that don’t contain any CTTCTT region in any of their exons.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#get_genelist_from_file-class_method" title="get_genelist_from_file (class method)">.<strong>get_genelist_from_file</strong>(filename)  &#x21d2; Array&lt;String&gt; </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Class method to read genelist file and create an array withall the genes.</p>
</div></span>
  
</li>

      
    </ul>
  



  <div id="class_method_details" class="method_details_list">
    <h2>Class Method Details</h2>

    
      <div class="method_details first">
  <h3 class="signature first" id="fetch-class_method">
  
    .<strong>fetch</strong>(url:, headers: {accept: &quot;*/*&quot;}, user: &quot;&quot;, pass: &quot;&quot;)  &#x21d2; <tt>String</tt>, <tt>false</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Class method fetch to access an URL via code, donated by Mark Wilkinson.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>url</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>URL</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>headers</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>{accept: &quot;*/*&quot;}</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>headers</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>user</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>&quot;&quot;</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>username, default “”</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>pass</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>&quot;&quot;</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>password, default “”</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>String</tt>, <tt>false</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the contents of the URL, or false if an exception occured</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'FileMaster.rb', line 56</span>

<span class='kw'>def</span> <span class='kw'>self</span><span class='period'>.</span><span class='id identifier rubyid_fetch'>fetch</span><span class='lparen'>(</span><span class='label'>url:</span><span class='comma'>,</span> <span class='label'>headers:</span> <span class='lbrace'>{</span><span class='label'>accept:</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>*/*</span><span class='tstring_end'>&quot;</span></span><span class='rbrace'>}</span><span class='comma'>,</span> <span class='label'>user:</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='label'>pass:</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>

    <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='const'>RestClient</span><span class='op'>::</span><span class='const'>Request</span><span class='period'>.</span><span class='id identifier rubyid_execute'>execute</span><span class='lparen'>(</span><span class='lbrace'>{</span>
        <span class='label'>method:</span> <span class='symbol'>:get</span><span class='comma'>,</span>
        <span class='label'>url:</span> <span class='id identifier rubyid_url'>url</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span><span class='comma'>,</span>
        <span class='label'>user:</span> <span class='id identifier rubyid_user'>user</span><span class='comma'>,</span>
        <span class='label'>password:</span> <span class='id identifier rubyid_pass'>pass</span><span class='comma'>,</span>
        <span class='label'>headers:</span> <span class='id identifier rubyid_headers'>headers</span><span class='rbrace'>}</span><span class='rparen'>)</span>
    <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span>
    
    <span class='kw'>rescue</span> <span class='const'>RestClient</span><span class='op'>::</span><span class='const'>ExceptionWithResponse</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span>
        <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
        <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='kw'>false</span>
        <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span>  
    
    <span class='kw'>rescue</span> <span class='const'>RestClient</span><span class='op'>::</span><span class='const'>Exception</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span>
        <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
        <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='kw'>false</span>
        <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span>  

    <span class='kw'>rescue</span> <span class='const'>Exception</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span>
        <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
        <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='kw'>false</span>
        <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span> 

<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="generate_gff3_report-class_method">
  
    .<strong>generate_gff3_report</strong>(seq_objects)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Class method to write GFF3 report with all the CTTCTT regions found in the genes from the list.  The coordinates are relative to the gene.</p>

<p>Creates a GFF3 report file of the CTTCTT regions with gene related coordinates</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>seq_objects</span>
      
      
        <span class='type'>(<tt>Array&lt;Object&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the list of all Seq objects as an Array</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'FileMaster.rb', line 113</span>

<span class='kw'>def</span> <span class='kw'>self</span><span class='period'>.</span><span class='id identifier rubyid_generate_gff3_report'>generate_gff3_report</span><span class='lparen'>(</span><span class='id identifier rubyid_seq_objects'>seq_objects</span><span class='rparen'>)</span>

    <span class='comment'># Common variables
</span>    <span class='id identifier rubyid_source'>source</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>Ruby</span><span class='tstring_end'>&quot;</span></span>
    <span class='id identifier rubyid_type'>type</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT_region</span><span class='tstring_end'>&quot;</span></span>
    <span class='id identifier rubyid_score'>score</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
    <span class='id identifier rubyid_phase'>phase</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
    
    <span class='comment'># Write file
</span>    <span class='id identifier rubyid_open'>open</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>GFF_genes.gff3</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>w</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_f'>f</span><span class='op'>|</span>

        <span class='comment'># Header
</span>        <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>##gff-version 3</span><span class='tstring_end'>&quot;</span></span> 

        <span class='id identifier rubyid_seq_objects'>seq_objects</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_seq_obj'>seq_obj</span><span class='op'>|</span> <span class='comment'># For each Seq object
</span>            <span class='id identifier rubyid_seqid'>seqid</span> <span class='op'>=</span> <span class='id identifier rubyid_seq_obj'>seq_obj</span><span class='period'>.</span><span class='id identifier rubyid_id'>id</span>        <span class='comment'># Get gene id
</span>            <span class='id identifier rubyid_biosequence'>biosequence</span> <span class='op'>=</span> <span class='id identifier rubyid_seq_obj'>seq_obj</span><span class='period'>.</span><span class='id identifier rubyid_sequence'>sequence</span>  <span class='comment'># Get the sequence attribute type Bio::Sequence object
</span>            <span class='id identifier rubyid_biosequence'>biosequence</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>
                <span class='kw'>next</span> <span class='kw'>unless</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT_region</span><span class='tstring_end'>&quot;</span></span> <span class='comment'># For each CTTCTT_region feature
</span>
                <span class='comment'># Gene relative coordinates 
</span>                <span class='id identifier rubyid_location'>location</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_locations'>locations</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span> <span class='comment'># Get location as Bio::Location
</span>                <span class='id identifier rubyid_start_pos'>start_pos</span> <span class='op'>=</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span>
                <span class='id identifier rubyid_end_pos'>end_pos</span> <span class='op'>=</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_to'>to</span>

                <span class='comment'># Rest of variables
</span>                <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span>
                <span class='id identifier rubyid_cttctt_id'>cttctt_id</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT_id</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span>
                <span class='id identifier rubyid_attributes'>attributes</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>gene_id=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_seqid'>seqid</span><span class='embexpr_end'>}</span><span class='tstring_content'>; CTTCTT_id=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_cttctt_id'>cttctt_id</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span>

                <span class='comment'># Write tab separated GFF3 file
</span>                <span class='id identifier rubyid_columns'>columns</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='id identifier rubyid_seqid'>seqid</span><span class='comma'>,</span> <span class='id identifier rubyid_source'>source</span><span class='comma'>,</span> <span class='id identifier rubyid_type'>type</span><span class='comma'>,</span> <span class='id identifier rubyid_start_pos'>start_pos</span><span class='comma'>,</span> <span class='id identifier rubyid_end_pos'>end_pos</span><span class='comma'>,</span> <span class='id identifier rubyid_score'>score</span><span class='comma'>,</span> <span class='id identifier rubyid_strand'>strand</span><span class='comma'>,</span> <span class='id identifier rubyid_phase'>phase</span><span class='comma'>,</span> <span class='id identifier rubyid_attributes'>attributes</span><span class='rbracket'>]</span>
                <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_columns'>columns</span><span class='period'>.</span><span class='id identifier rubyid_join'>join</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>

            <span class='kw'>end</span>
        <span class='kw'>end</span>

    <span class='kw'>end</span>

<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="generate_gff3_report_chr-class_method">
  
    .<strong>generate_gff3_report_chr</strong>(seq_objects)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Class method to write GFF3 report with all the CTTCTT regions found in the genes from the list.  The coordinates are absolute as they take the chromosome as reference.</p>

<p>Creates a GFF3 report file of the CTTCTT regions with chromosome related coordinates.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>seq_objects</span>
      
      
        <span class='type'>(<tt>Array&lt;Object&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the list of all Seq objects as an Array.</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'FileMaster.rb', line 161</span>

<span class='kw'>def</span> <span class='kw'>self</span><span class='period'>.</span><span class='id identifier rubyid_generate_gff3_report_chr'>generate_gff3_report_chr</span><span class='lparen'>(</span><span class='id identifier rubyid_seq_objects'>seq_objects</span><span class='rparen'>)</span>

    <span class='comment'># Common variables
</span>    <span class='id identifier rubyid_source'>source</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>Ruby</span><span class='tstring_end'>&quot;</span></span>
    <span class='id identifier rubyid_type'>type</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT_region</span><span class='tstring_end'>&quot;</span></span>
    <span class='id identifier rubyid_score'>score</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
    <span class='id identifier rubyid_phase'>phase</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
    
    <span class='comment'># Write file
</span>    <span class='id identifier rubyid_open'>open</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>GFF_chr.gff3</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>w</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_f'>f</span><span class='op'>|</span>

        <span class='comment'># Header
</span>        <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>##gff-version 3</span><span class='tstring_end'>&quot;</span></span> 

        <span class='id identifier rubyid_seq_objects'>seq_objects</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_seq_obj'>seq_obj</span><span class='op'>|</span> <span class='comment'># For each Seq object
</span>            <span class='id identifier rubyid_chr'>chr</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>Chr</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_seq_obj'>seq_obj</span><span class='period'>.</span><span class='id identifier rubyid_chromosome'>chromosome</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='comment'># Get chromosome 
</span>            <span class='id identifier rubyid_biosequence'>biosequence</span> <span class='op'>=</span> <span class='id identifier rubyid_seq_obj'>seq_obj</span><span class='period'>.</span><span class='id identifier rubyid_sequence'>sequence</span> <span class='comment'># Get the sequence attribute type Bio::Sequence object
</span>            <span class='id identifier rubyid_biosequence'>biosequence</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>
                <span class='kw'>next</span> <span class='kw'>unless</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT_region</span><span class='tstring_end'>&quot;</span></span> <span class='comment'># For each CTTCTT_region feature
</span>                
                <span class='comment'># Chromosome relative coordinates 
</span>                <span class='id identifier rubyid_location'>location</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_locations'>locations</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span> <span class='comment'># Get location as Bio::Location
</span>                <span class='id identifier rubyid_start_pos'>start_pos</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_seq_obj'>seq_obj</span><span class='period'>.</span><span class='id identifier rubyid_start_gene_chr'>start_gene_chr</span> <span class='op'>+</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>-</span> <span class='int'>1</span><span class='rparen'>)</span>  <span class='comment'># -1 to add 1-base offset
</span>                <span class='id identifier rubyid_end_pos'>end_pos</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_seq_obj'>seq_obj</span><span class='period'>.</span><span class='id identifier rubyid_start_gene_chr'>start_gene_chr</span> <span class='op'>+</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_to'>to</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>-</span> <span class='int'>1</span><span class='rparen'>)</span> <span class='comment'># -1 to add 1-base offset
</span>                
                <span class='comment'># Rest of variables
</span>                <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span>
                <span class='id identifier rubyid_cttctt_id'>cttctt_id</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT_id</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span>
                <span class='id identifier rubyid_attributes'>attributes</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>gene_id=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_seq_obj'>seq_obj</span><span class='period'>.</span><span class='id identifier rubyid_id'>id</span><span class='embexpr_end'>}</span><span class='tstring_content'>; CTTCTT_id=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_cttctt_id'>cttctt_id</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span>

                <span class='comment'># Write tab separated GFF3 file
</span>                <span class='id identifier rubyid_columns'>columns</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='id identifier rubyid_chr'>chr</span><span class='comma'>,</span> <span class='id identifier rubyid_source'>source</span><span class='comma'>,</span> <span class='id identifier rubyid_type'>type</span><span class='comma'>,</span> <span class='id identifier rubyid_start_pos'>start_pos</span><span class='comma'>,</span> <span class='id identifier rubyid_end_pos'>end_pos</span><span class='comma'>,</span> <span class='id identifier rubyid_score'>score</span><span class='comma'>,</span> <span class='id identifier rubyid_strand'>strand</span><span class='comma'>,</span> <span class='id identifier rubyid_phase'>phase</span><span class='comma'>,</span> <span class='id identifier rubyid_attributes'>attributes</span><span class='rbracket'>]</span>
                <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_columns'>columns</span><span class='period'>.</span><span class='id identifier rubyid_join'>join</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>

            <span class='kw'>end</span>
        <span class='kw'>end</span>

    <span class='kw'>end</span>

<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="generate_no_report-class_method">
  
    .<strong>generate_no_report</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Class method to write report with all the genes that don’t contain any CTTCTT region in any of their exons</p>

<p>Creates the no_CTTCTT_genes_report</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'FileMaster.rb', line 88</span>

<span class='kw'>def</span> <span class='kw'>self</span><span class='period'>.</span><span class='id identifier rubyid_generate_no_report'>generate_no_report</span>

    <span class='comment'># Save the information in local variables
</span>    <span class='id identifier rubyid_no_genes'>no_genes</span> <span class='op'>=</span> <span class='const'>Seq</span><span class='period'>.</span><span class='id identifier rubyid_get_no_genes'>get_no_genes</span>
    <span class='id identifier rubyid_no_genes_count'>no_genes_count</span> <span class='op'>=</span> <span class='id identifier rubyid_no_genes'>no_genes</span><span class='period'>.</span><span class='id identifier rubyid_length'>length</span><span class='lparen'>(</span><span class='rparen'>)</span>
    <span class='id identifier rubyid_total_genes'>total_genes</span> <span class='op'>=</span> <span class='cvar'>@@genelist</span><span class='period'>.</span><span class='id identifier rubyid_length'>length</span><span class='lparen'>(</span><span class='rparen'>)</span>

    <span class='comment'># Write file
</span>    <span class='id identifier rubyid_open'>open</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>no_CTTCTT_genes_report.txt</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>w</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_f'>f</span><span class='op'>|</span>
        <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_total_genes'>total_genes</span><span class='embexpr_end'>}</span><span class='tstring_content'> were evaluated</span><span class='tstring_end'>&quot;</span></span>
        <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_no_genes_count'>no_genes_count</span><span class='embexpr_end'>}</span><span class='tstring_content'> genes do not have any CTTCTT region in any of their exons:</span><span class='tstring_end'>&quot;</span></span>
        <span class='id identifier rubyid_no_genes'>no_genes</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_gene'>gene</span><span class='op'>|</span>
            <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_gene'>gene</span>
        <span class='kw'>end</span>
    <span class='kw'>end</span>

<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="get_genelist_from_file-class_method">
  
    .<strong>get_genelist_from_file</strong>(filename)  &#x21d2; <tt>Array&lt;String&gt;</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Class method to read genelist file and create an array withall the genes</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>filename</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the name of the file cointaiing the gene list as a String</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Array&lt;String&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the gene list</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'FileMaster.rb', line 27</span>

<span class='kw'>def</span> <span class='kw'>self</span><span class='period'>.</span><span class='id identifier rubyid_get_genelist_from_file'>get_genelist_from_file</span><span class='lparen'>(</span><span class='id identifier rubyid_filename'>filename</span><span class='rparen'>)</span>

    <span class='kw'>if</span> <span class='cvar'>@@genelist</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='comment'># Only if genelist has not been previously defined
</span>
        <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_open'>open</span><span class='lparen'>(</span><span class='id identifier rubyid_filename'>filename</span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>r</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_f'>f</span><span class='op'>|</span>

            <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_readlines'>readlines</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_l'>l</span><span class='op'>|</span>
                <span class='id identifier rubyid_geneid'>geneid</span> <span class='op'>=</span> <span class='id identifier rubyid_l'>l</span><span class='period'>.</span><span class='id identifier rubyid_chop'>chop</span>                <span class='comment'># Remove final \n
</span>                <span class='cvar'>@@genelist</span><span class='period'>.</span><span class='id identifier rubyid_push'>push</span><span class='lparen'>(</span><span class='id identifier rubyid_geneid'>geneid</span><span class='period'>.</span><span class='id identifier rubyid_upcase'>upcase</span><span class='rparen'>)</span> <span class='comment'># Add genes in all caps
</span>            <span class='kw'>end</span>

        <span class='kw'>end</span>

    <span class='kw'>end</span>

    <span class='kw'>return</span> <span class='cvar'>@@genelist</span>

<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
  </div>

</div>

      <div id="footer">
  Generated on Tue Dec 20 18:03:56 2022 by
  <a href="https://yardoc.org" title="Yay! A Ruby Documentation Tool" target="_parent">yard</a>
  0.9.28 (ruby-3.1.2).
</div>

    </div>
  </body>
</html>