<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>
  Class: Seq
  
    &mdash; Documentation by YARD 0.9.28
  
</title>

  <link rel="stylesheet" href="css/style.css" type="text/css" />

  <link rel="stylesheet" href="css/common.css" type="text/css" />

<script type="text/javascript">
  pathId = "Seq";
  relpath = '';
</script>


  <script type="text/javascript" charset="utf-8" src="js/jquery.js"></script>

  <script type="text/javascript" charset="utf-8" src="js/app.js"></script>


  </head>
  <body>
    <div class="nav_wrap">
      <iframe id="nav" src="class_list.html?1"></iframe>
      <div id="resizer"></div>
    </div>

    <div id="main" tabindex="-1">
      <div id="header">
        <div id="menu">
  
    <a href="_index.html">Index (S)</a> &raquo;
    
    
    <span class="title">Seq</span>
  
</div>

        <div id="search">
  
    <a class="full_list_link" id="class_list_link"
        href="class_list.html">

        <svg width="24" height="24">
          <rect x="0" y="4" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="12" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="20" width="24" height="4" rx="1" ry="1"></rect>
        </svg>
    </a>
  
</div>
        <div class="clear"></div>
      </div>

      <div id="content"><h1>Class: Seq
  
  
  
</h1>
<div class="box_info">
  
  <dl>
    <dt>Inherits:</dt>
    <dd>
      <span class="inheritName">Object</span>
      
        <ul class="fullTree">
          <li>Object</li>
          
            <li class="next">Seq</li>
          
        </ul>
        <a href="#" class="inheritanceTree">show all</a>
      
    </dd>
  </dl>
  

  
  
  
  
  

  

  
  <dl>
    <dt>Defined in:</dt>
    <dd>Seq.rb</dd>
  </dl>
  
</div>

<h2>Overview</h2><div class="docstring">
  <div class="discussion">
    
<h2 id="label-Seq">Seq</h2>

<p>Class to represent all of the information associated with a gene</p>

<p>Contains the functions to get the no_CTTCTT genes list, to get the bio records of embl and and sequence,  to get the the chromosome number and the chromosome related coordinates, to add new features to the bio sequence, to find the matches of the sequence in the gene exons.</p>

<h2 id="label-Summary">Summary</h2>

<p>This can be used to represent all the information associated with a gene</p>


  </div>
</div>
<div class="tags">
  

</div>
  
    <h2>
      Constant Summary
      <small><a href="#" class="constants_summary_toggle">collapse</a></small>
    </h2>

    <dl class="constants">
      
        <dt id="no_cttctt_genes-classvariable" class="">@@no_cttctt_genes =
          <div class="docstring">
  <div class="discussion">
    
<p>Class variable [Array&lt;String&gt;]  Contains the genes id of the genes that don’t contain any CTTCTT repeats in any of their exons</p>


  </div>
</div>
<div class="tags">
  

</div>
        </dt>
        <dd><pre class="code"><span class='lbracket'>[</span><span class='rbracket'>]</span></pre></dd>
      
    </dl>
  




  <h2>Instance Attribute Summary <small><a href="#" class="summary_toggle">collapse</a></small></h2>
  <ul class="summary">
    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#chromosome-instance_method" title="#chromosome (instance method)">#<strong>chromosome</strong>  &#x21d2; String </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Get/Set the chromosome number in which the gene is located.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#embl-instance_method" title="#embl (instance method)">#<strong>embl</strong>  &#x21d2; Bio::EMBL </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Get/Set the embl associated with a gene id.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#end_gene_chr-instance_method" title="#end_gene_chr (instance method)">#<strong>end_gene_chr</strong>  &#x21d2; Integer </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Get/Set the ending position of the chromosome in the genome.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#id-instance_method" title="#id (instance method)">#<strong>id</strong>  &#x21d2; String </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Get/Set the id of the gene sequence.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#sequence-instance_method" title="#sequence (instance method)">#<strong>sequence</strong>  &#x21d2; Bio::Sequence </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Get/Set the biosequence extracted from the embl.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#start_gene_chr-instance_method" title="#start_gene_chr (instance method)">#<strong>start_gene_chr</strong>  &#x21d2; Integer </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Get/Set the starting position of the chromosome in the genome.</p>
</div></span>
  
</li>

    
  </ul>




  
    <h2>
      Class Method Summary
      <small><a href="#" class="summary_toggle">collapse</a></small>
    </h2>

    <ul class="summary">
      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#get_no_genes-class_method" title="get_no_genes (class method)">.<strong>get_no_genes</strong>  &#x21d2; Array&lt;String&gt; </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Function to get the the class variable @@no_cttctt_genes.</p>
</div></span>
  
</li>

      
    </ul>
  
    <h2>
      Instance Method Summary
      <small><a href="#" class="summary_toggle">collapse</a></small>
    </h2>

    <ul class="summary">
      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#add_new_feature-instance_method" title="#add_new_feature (instance method)">#<strong>add_new_feature</strong>(strand, position_s, index)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Create new CTTCTT_region feature and add it to the Bio::Sequence object @sequence.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#get_bio-instance_method" title="#get_bio (instance method)">#<strong>get_bio</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Obtains information for the current genes.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#get_coords_chr-instance_method" title="#get_coords_chr (instance method)">#<strong>get_coords_chr</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Obtains the chromose number and the chromosome relative coordinates of the gene.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#initialize-instance_method" title="#initialize (instance method)">#<strong>initialize</strong>(id:)  &#x21d2; sequence </a>
    

    
  </span>
  
  
    <span class="note title constructor">constructor</span>
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Inizialize method to create a new instance of Seq.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#match_exons-instance_method" title="#match_exons (instance method)">#<strong>match_exons</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Finds every CTTCTT region in the exons of the current gene on both the + and - strands,  adds a new feature to its @sequence and updates @@no_cttctt_genes array.</p>
</div></span>
  
</li>

      
    </ul>
  

<div id="constructor_details" class="method_details_list">
  <h2>Constructor Details</h2>
  
    <div class="method_details first">
  <h3 class="signature first" id="initialize-instance_method">
  
    #<strong>initialize</strong>(id:)  &#x21d2; <tt><span class='object_link'><a href="#sequence-instance_method" title="Seq#sequence (method)">sequence</a></span></tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Inizialize method to create a new instance of Seq. Called with new</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>id</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the id of the gene as a String</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


59
60
61
62
63</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Seq.rb', line 59</span>

<span class='kw'>def</span> <span class='id identifier rubyid_initialize'>initialize</span><span class='lparen'>(</span><span class='label'>id:</span><span class='rparen'>)</span>

    <span class='ivar'>@id</span> <span class='op'>=</span> <span class='id identifier rubyid_id'>id</span>

<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
  
</div>

  <div id="instance_attr_details" class="attr_details">
    <h2>Instance Attribute Details</h2>
    
      
      <span id="chromosome=-instance_method"></span>
      <div class="method_details first">
  <h3 class="signature first" id="chromosome-instance_method">
  
    #<strong>chromosome</strong>  &#x21d2; <tt>String</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Get/Set the chromosome number in which the gene is located</p>


  </div>
</div>
<div class="tags">
  
<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The chromosome number</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


36
37
38</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Seq.rb', line 36</span>

<span class='kw'>def</span> <span class='id identifier rubyid_chromosome'>chromosome</span>
  <span class='ivar'>@chromosome</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="embl=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="embl-instance_method">
  
    #<strong>embl</strong>  &#x21d2; <tt>Bio::EMBL</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Get/Set the embl associated with a gene id</p>


  </div>
</div>
<div class="tags">
  
<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Bio::EMBL</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The bio embl record</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


26
27
28</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Seq.rb', line 26</span>

<span class='kw'>def</span> <span class='id identifier rubyid_embl'>embl</span>
  <span class='ivar'>@embl</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="end_gene_chr=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="end_gene_chr-instance_method">
  
    #<strong>end_gene_chr</strong>  &#x21d2; <tt>Integer</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Get/Set the ending position of the chromosome in the genome</p>


  </div>
</div>
<div class="tags">
  
<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Integer</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The chromosome ending position</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


46
47
48</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Seq.rb', line 46</span>

<span class='kw'>def</span> <span class='id identifier rubyid_end_gene_chr'>end_gene_chr</span>
  <span class='ivar'>@end_gene_chr</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="id=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="id-instance_method">
  
    #<strong>id</strong>  &#x21d2; <tt>String</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Get/Set the id of the gene sequence</p>


  </div>
</div>
<div class="tags">
  
<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The gene id</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


21
22
23</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Seq.rb', line 21</span>

<span class='kw'>def</span> <span class='id identifier rubyid_id'>id</span>
  <span class='ivar'>@id</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="sequence=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="sequence-instance_method">
  
    #<strong>sequence</strong>  &#x21d2; <tt>Bio::Sequence</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Get/Set the biosequence extracted from the embl</p>


  </div>
</div>
<div class="tags">
  
<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Bio::Sequence</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The bio sequence record</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


31
32
33</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Seq.rb', line 31</span>

<span class='kw'>def</span> <span class='id identifier rubyid_sequence'>sequence</span>
  <span class='ivar'>@sequence</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="start_gene_chr=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="start_gene_chr-instance_method">
  
    #<strong>start_gene_chr</strong>  &#x21d2; <tt>Integer</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Get/Set the starting position of the chromosome in the genome</p>


  </div>
</div>
<div class="tags">
  
<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Integer</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The chromosome starting position</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


41
42
43</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Seq.rb', line 41</span>

<span class='kw'>def</span> <span class='id identifier rubyid_start_gene_chr'>start_gene_chr</span>
  <span class='ivar'>@start_gene_chr</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
  </div>


  <div id="class_method_details" class="method_details_list">
    <h2>Class Method Details</h2>

    
      <div class="method_details first">
  <h3 class="signature first" id="get_no_genes-class_method">
  
    .<strong>get_no_genes</strong>  &#x21d2; <tt>Array&lt;String&gt;</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Function to get the the class variable @@no_cttctt_genes</p>


  </div>
</div>
<div class="tags">
  
<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Array&lt;String&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The Array with all of the gene id that don’t have any cttctt repeat in their exons</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


70
71
72
73
74</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Seq.rb', line 70</span>

<span class='kw'>def</span> <span class='kw'>self</span><span class='period'>.</span><span class='id identifier rubyid_get_no_genes'>get_no_genes</span>

    <span class='kw'>return</span> <span class='cvar'>@@no_cttctt_genes</span><span class='period'>.</span><span class='id identifier rubyid_uniq'>uniq</span>

<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
  </div>

  <div id="instance_method_details" class="method_details_list">
    <h2>Instance Method Details</h2>

    
      <div class="method_details first">
  <h3 class="signature first" id="add_new_feature-instance_method">
  
    #<strong>add_new_feature</strong>(strand, position_s, index)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Create new CTTCTT_region feature and add it to the Bio::Sequence object @sequence</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>strand</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>“+” for forward strand, “-” for the reverse strand</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>position_s</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>string that stores the repeat position as “start_pos..end_pos”</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>index</span>
      
      
        <span class='type'>(<tt>Integer</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>to keep a record of all of the CTTCTT regions in the sequence</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


123
124
125
126
127
128
129
130
131
132
133
134</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Seq.rb', line 123</span>

<span class='kw'>def</span> <span class='id identifier rubyid_add_new_feature'>add_new_feature</span><span class='lparen'>(</span><span class='id identifier rubyid_strand'>strand</span><span class='comma'>,</span> <span class='id identifier rubyid_position_s'>position_s</span><span class='comma'>,</span> <span class='id identifier rubyid_index'>index</span><span class='rparen'>)</span>

    <span class='id identifier rubyid_new_feature'>new_feature</span> <span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span>    <span class='comment'># Creating new feature 
</span>        <span class='id identifier rubyid_feature'>feature</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT_region</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span>     <span class='comment'># Feature name
</span>        <span class='id identifier rubyid_position'>position</span> <span class='op'>=</span> <span class='id identifier rubyid_position_s'>position_s</span><span class='comma'>,</span>         <span class='comment'># Position with Bio::Location format: &quot;start_pos..end_pos&quot;
</span>        <span class='id identifier rubyid_qualifiers'>qualifiers</span> <span class='op'>=</span> <span class='lbracket'>[</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>sequence</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span><span class='comma'>,</span>
                       <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_strand'>strand</span><span class='rparen'>)</span><span class='comma'>,</span>
                       <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>CTTCTT_id</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='ivar'>@id</span><span class='embexpr_end'>}</span><span class='tstring_content'>.CTTCTT.</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_strand'>strand</span><span class='embexpr_end'>}</span><span class='tstring_content'>.</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_index'>index</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> <span class='rbracket'>]</span> <span class='rparen'>)</span>

    <span class='ivar'>@sequence</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_new_feature'>new_feature</span>  <span class='comment'># Add new_feature to Bio::Sequence object @sequence
</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="get_bio-instance_method">
  
    #<strong>get_bio</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Obtains information for the current genes.</p>

<p>Takes the instance atribute @id and  saves the [Bio::EMBL] and [Bio::Sequence] created objects in its respective instance atributes @embl and @sequence</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


82
83
84
85
86
87
88
89
90
91</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Seq.rb', line 82</span>

<span class='kw'>def</span> <span class='id identifier rubyid_get_bio'>get_bio</span> <span class='comment'>#Obtaining gene information for the current gene (gene id)
</span>
    <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='const'>FileMaster</span><span class='period'>.</span><span class='id identifier rubyid_fetch'>fetch</span><span class='lparen'>(</span><span class='label'>url:</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&amp;format=embl&amp;id=</span><span class='embexpr_beg'>#{</span><span class='ivar'>@id</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>  <span class='comment'># Fetch EMBL file
</span>
    <span class='kw'>if</span> <span class='id identifier rubyid_response'>response</span> 
        <span class='ivar'>@embl</span> <span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>EMBL</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='id identifier rubyid_response'>response</span><span class='period'>.</span><span class='id identifier rubyid_body'>body</span><span class='rparen'>)</span> <span class='comment'># Transform the EMBL file into a Bio::EMBL object and save it into @embl
</span>        <span class='ivar'>@sequence</span> <span class='op'>=</span> <span class='ivar'>@embl</span><span class='period'>.</span><span class='id identifier rubyid_to_biosequence'>to_biosequence</span>  <span class='comment'># Transforming the Bio::EMBL into Bio::Sequence object and save it into @sequence
</span>    <span class='kw'>end</span>

<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="get_coords_chr-instance_method">
  
    #<strong>get_coords_chr</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Obtains the chromose number and the chromosome relative coordinates of the gene</p>

<p>Takes the @embl instance variable and  saves the chromosome number and chromosome relative coordinates of the gene  on its respective instance attributes @chromosome, @start_gene_chr and @end_gene_chr</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


100
101
102
103
104
105
106
107
108
109
110
111
112
113
114</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Seq.rb', line 100</span>

<span class='kw'>def</span> <span class='id identifier rubyid_get_coords_chr'>get_coords_chr</span>

    <span class='comment'># chromosome information is stored in the accession of the Bio::EMBL object. 
</span>    <span class='comment'># Accesion format: chromosome:TAIR10:5:22038165:22039568:1
</span>    <span class='id identifier rubyid_acc'>acc</span> <span class='op'>=</span> <span class='ivar'>@embl</span><span class='period'>.</span><span class='id identifier rubyid_accession'>accession</span><span class='period'>.</span><span class='id identifier rubyid_split'>split</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>:</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> 
    <span class='id identifier rubyid_chr'>chr</span> <span class='op'>=</span> <span class='id identifier rubyid_acc'>acc</span><span class='lbracket'>[</span><span class='int'>2</span><span class='rbracket'>]</span>
    <span class='id identifier rubyid_start_chr'>start_chr</span> <span class='op'>=</span> <span class='id identifier rubyid_acc'>acc</span><span class='lbracket'>[</span><span class='int'>3</span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span>
    <span class='id identifier rubyid_end_chr'>end_chr</span> <span class='op'>=</span> <span class='id identifier rubyid_acc'>acc</span><span class='lbracket'>[</span><span class='int'>4</span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span>

    <span class='comment'># Save chromosome related information into its respective instance variables
</span>    <span class='ivar'>@chromosome</span> <span class='op'>=</span> <span class='id identifier rubyid_chr'>chr</span>
    <span class='ivar'>@start_gene_chr</span> <span class='op'>=</span> <span class='id identifier rubyid_start_chr'>start_chr</span>
    <span class='ivar'>@end_gene_chr</span> <span class='op'>=</span> <span class='id identifier rubyid_end_chr'>end_chr</span>

<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="match_exons-instance_method">
  
    #<strong>match_exons</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Finds every CTTCTT region in the exons of the current gene on both the + and - strands,  adds a new feature to its @sequence and updates @@no_cttctt_genes array</p>

<p>Takes the @embl [Bio::EMBL] instance variable and  creates a new CTTCTT_region feature for its @sequence [Bio::Sequence] If the CTTCTT sequence is not in any of the exons, the @id [String] is added to @@no_cttctt_genes [Array]</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Seq.rb', line 144</span>

<span class='kw'>def</span> <span class='id identifier rubyid_match_exons'>match_exons</span>

    <span class='comment'># Initializing local variables
</span>    <span class='id identifier rubyid_match_positions_direct'>match_positions_direct</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span>
    <span class='id identifier rubyid_match_positions_reverse'>match_positions_reverse</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span>

    <span class='comment'># Iterating over the features of the Bio::EMBL object
</span>    <span class='ivar'>@embl</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>

        <span class='comment'># Getting the exon nucleotide sequence
</span>        <span class='kw'>next</span> <span class='kw'>unless</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>exon</span><span class='tstring_end'>&quot;</span></span>            <span class='comment'># Checking only exons
</span>        <span class='id identifier rubyid_location'>location</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_locations'>locations</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span>                  <span class='comment'># Position of the exon relative to the gene as Bio::Location object
</span>        <span class='id identifier rubyid_start_pos'>start_pos</span><span class='comma'>,</span> <span class='id identifier rubyid_end_pos'>end_pos</span> <span class='op'>=</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span><span class='comma'>,</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_to'>to</span>  <span class='comment'># Saving the start and end position of the exon
</span>        <span class='kw'>next</span> <span class='kw'>unless</span> <span class='id identifier rubyid_end_pos'>end_pos</span><span class='period'>.</span><span class='id identifier rubyid_between?'>between?</span><span class='lparen'>(</span><span class='int'>0</span><span class='comma'>,</span><span class='ivar'>@embl</span><span class='period'>.</span><span class='id identifier rubyid_seqlen'>seqlen</span><span class='rparen'>)</span>     <span class='comment'># Next if exon is out of sequence range 
</span>        <span class='id identifier rubyid_exon_seq'>exon_seq</span> <span class='op'>=</span> <span class='ivar'>@sequence</span><span class='period'>.</span><span class='id identifier rubyid_subseq'>subseq</span><span class='lparen'>(</span><span class='id identifier rubyid_start_pos'>start_pos</span><span class='comma'>,</span><span class='id identifier rubyid_end_pos'>end_pos</span><span class='rparen'>)</span>   <span class='comment'># Getting the sequence of the exon
</span>        <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span>                          <span class='comment'># Next if exon sequence is empty
</span>        
        <span class='comment'># Finding a CTTCTT region when exon is in the + strand
</span>        <span class='kw'>if</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='int'>1</span>

            <span class='comment'># Match region
</span>            <span class='id identifier rubyid_match_start_rel'>match_start_rel</span> <span class='op'>=</span> <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='period'>.</span><span class='id identifier rubyid_enum_for'>enum_for</span><span class='lparen'>(</span><span class='symbol'>:scan</span><span class='comma'>,</span> <span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>(?=(cttctt))</span><span class='regexp_end'>/i</span></span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_map'>map</span> <span class='lbrace'>{</span><span class='const'>Regexp</span><span class='period'>.</span><span class='id identifier rubyid_last_match'>last_match</span><span class='period'>.</span><span class='id identifier rubyid_begin'>begin</span><span class='lparen'>(</span><span class='int'>0</span><span class='rparen'>)</span><span class='rbrace'>}</span> <span class='comment'># Match and get position
</span>            <span class='comment'># https://stackoverflow.com/questions/5241653/ruby-regex-match-and-get-positions-of
</span>            <span class='comment'># i ignores case while matching alphabets https://learnbyexample.github.io/Ruby_Regexp/modifiers.html 
</span>            <span class='comment'># ?= Lookahead to find overlapping regions https://learnbyexample.github.io/Ruby_Regexp/lookarounds.html 
</span>            <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_match_start_rel'>match_start_rel</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span>
            
            <span class='comment'># Saving gene relative coordinates of the CTTCTT region
</span>            <span class='id identifier rubyid_match_start_rel'>match_start_rel</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_start'>start</span><span class='op'>|</span>                    <span class='comment'># For each match found
</span>                <span class='id identifier rubyid_match_start'>match_start</span> <span class='op'>=</span> <span class='id identifier rubyid_start'>start</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>+</span> <span class='id identifier rubyid_start_pos'>start_pos</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span>      <span class='comment'># Start of the region in the exon + start of the exon in the gene
</span>                <span class='id identifier rubyid_match_end'>match_end</span> <span class='op'>=</span> <span class='id identifier rubyid_start'>start</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>+</span> <span class='id identifier rubyid_start_pos'>start_pos</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>+</span> <span class='int'>5</span>    <span class='comment'># 5 is the length -1 of the region (1-base offset)
</span>                <span class='id identifier rubyid_position_s'>position_s</span> <span class='op'>=</span>  <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_match_start'>match_start</span><span class='embexpr_end'>}</span><span class='tstring_content'>..</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_match_end'>match_end</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span>   <span class='comment'># Save position with Bio::Location format
</span>                <span class='id identifier rubyid_match_positions_direct'>match_positions_direct</span><span class='period'>.</span><span class='id identifier rubyid_push'>push</span><span class='lparen'>(</span><span class='id identifier rubyid_position_s'>position_s</span><span class='rparen'>)</span>        <span class='comment'># Add it to the matching positions direct array
</span>            <span class='kw'>end</span>

        <span class='comment'># Finding a CTTCTT region when exon is in the - strand    
</span>        <span class='kw'>elsif</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='op'>-</span><span class='int'>1</span>

            <span class='comment'># Match region
</span>            <span class='id identifier rubyid_match_start_rel'>match_start_rel</span> <span class='op'>=</span> <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='period'>.</span><span class='id identifier rubyid_enum_for'>enum_for</span><span class='lparen'>(</span><span class='symbol'>:scan</span><span class='comma'>,</span> <span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>(?=(aagaag))</span><span class='regexp_end'>/i</span></span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_map'>map</span> <span class='lbrace'>{</span><span class='const'>Regexp</span><span class='period'>.</span><span class='id identifier rubyid_last_match'>last_match</span><span class='period'>.</span><span class='id identifier rubyid_begin'>begin</span><span class='lparen'>(</span><span class='int'>0</span><span class='rparen'>)</span><span class='rbrace'>}</span> <span class='comment'># aagaag is the reverse complement of cttctt
</span>            <span class='comment'># https://stackoverflow.com/questions/5241653/ruby-regex-match-and-get-positions-of
</span>            <span class='comment'># i ignores case while matching alphabets https://learnbyexample.github.io/Ruby_Regexp/modifiers.html 
</span>            <span class='comment'># ?= Lookahead to find overlapping regions https://learnbyexample.github.io/Ruby_Regexp/lookarounds.html 
</span>            <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_match_start_rel'>match_start_rel</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> 
            
            <span class='comment'># Saving gene relative coordinates of the CTTCTT region
</span>            <span class='id identifier rubyid_match_start_rel'>match_start_rel</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_start'>start</span><span class='op'>|</span>                    <span class='comment'># For each match found
</span>                <span class='id identifier rubyid_match_start'>match_start</span> <span class='op'>=</span> <span class='id identifier rubyid_start'>start</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>+</span> <span class='id identifier rubyid_start_pos'>start_pos</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span>      <span class='comment'># Start of the region in the exon + start of the exon in the gene
</span>                <span class='id identifier rubyid_match_end'>match_end</span> <span class='op'>=</span> <span class='id identifier rubyid_start'>start</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>+</span> <span class='id identifier rubyid_start_pos'>start_pos</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>+</span> <span class='int'>5</span>    <span class='comment'># 5 is the length -1 of the region (1-base offset)
</span>                <span class='id identifier rubyid_position_s'>position_s</span> <span class='op'>=</span>  <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_match_start'>match_start</span><span class='embexpr_end'>}</span><span class='tstring_content'>..</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_match_end'>match_end</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span>   <span class='comment'># Save position with Bio::Location format
</span>                <span class='id identifier rubyid_match_positions_reverse'>match_positions_reverse</span><span class='period'>.</span><span class='id identifier rubyid_push'>push</span><span class='lparen'>(</span><span class='id identifier rubyid_position_s'>position_s</span><span class='rparen'>)</span>       <span class='comment'># Add it to the matching positions reverse array
</span>            <span class='kw'>end</span>

        <span class='kw'>end</span>

    <span class='kw'>end</span>
    
    <span class='comment'># Eliminate repeated matchs if any
</span>    <span class='id identifier rubyid_match_positions_direct'>match_positions_direct</span> <span class='op'>=</span> <span class='id identifier rubyid_match_positions_direct'>match_positions_direct</span><span class='period'>.</span><span class='id identifier rubyid_uniq'>uniq</span>
    <span class='id identifier rubyid_match_positions_reverse'>match_positions_reverse</span> <span class='op'>=</span> <span class='id identifier rubyid_match_positions_reverse'>match_positions_reverse</span><span class='period'>.</span><span class='id identifier rubyid_uniq'>uniq</span>

    
    <span class='kw'>if</span> <span class='id identifier rubyid_match_positions_direct'>match_positions_direct</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='op'>&amp;&amp;</span> <span class='id identifier rubyid_match_positions_reverse'>match_positions_reverse</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='comment'># If no match was found in any of the exons...
</span>        <span class='cvar'>@@no_cttctt_genes</span><span class='period'>.</span><span class='id identifier rubyid_push'>push</span><span class='lparen'>(</span><span class='ivar'>@id</span><span class='rparen'>)</span> <span class='comment'># ...Add @id to the @@no_cttctt_genes array
</span>    <span class='kw'>end</span>

    <span class='comment'># Add new CTTCTT_region features for the direct strand matches
</span>    <span class='id identifier rubyid_index'>index</span> <span class='op'>=</span> <span class='int'>1</span>
    <span class='id identifier rubyid_match_positions_direct'>match_positions_direct</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_position'>position</span><span class='op'>|</span>
        <span class='id identifier rubyid_strand_s'>strand_s</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>+</span><span class='tstring_end'>&quot;</span></span>
        <span class='id identifier rubyid_add_new_feature'>add_new_feature</span><span class='lparen'>(</span><span class='id identifier rubyid_strand_s'>strand_s</span><span class='comma'>,</span> <span class='id identifier rubyid_position'>position</span><span class='comma'>,</span> <span class='id identifier rubyid_index'>index</span><span class='rparen'>)</span>
        <span class='id identifier rubyid_index'>index</span><span class='op'>+=</span><span class='int'>1</span>
    <span class='kw'>end</span>

    <span class='comment'># Add new CTTCTT_region features for the reverse strand matches
</span>    <span class='id identifier rubyid_index'>index</span> <span class='op'>=</span> <span class='int'>1</span>
    <span class='id identifier rubyid_match_positions_reverse'>match_positions_reverse</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_position'>position</span><span class='op'>|</span>
        <span class='id identifier rubyid_strand_s'>strand_s</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>-</span><span class='tstring_end'>&quot;</span></span>
        <span class='id identifier rubyid_add_new_feature'>add_new_feature</span><span class='lparen'>(</span><span class='id identifier rubyid_strand_s'>strand_s</span><span class='comma'>,</span> <span class='id identifier rubyid_position'>position</span><span class='comma'>,</span> <span class='id identifier rubyid_index'>index</span><span class='rparen'>)</span>
        <span class='id identifier rubyid_index'>index</span><span class='op'>+=</span><span class='int'>1</span>
    <span class='kw'>end</span>

<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
  </div>

</div>

      <div id="footer">
  Generated on Tue Dec 20 18:04:04 2022 by
  <a href="https://yardoc.org" title="Yay! A Ruby Documentation Tool" target="_parent">yard</a>
  0.9.28 (ruby-3.1.2).
</div>

    </div>
  </body>
</html>